duplex_http_call.go

```diff
@@ -257,15 +263,15 @@ func (d *duplexHTTPCall) makeRequest() {
 			err = NewError(CodeUnavailable, err)
 		}
 		d.responseErr = err
-		d.requestBodyReader.CloseWithError(io.EOF)
+		d.requestBodyWriter.Close()
```

Comment:
Why is this changed? Since it was using `CloseWithError` and not simply `Close`, this would not be the source of `io.ErrPipeClosed` issues. In fact, I believe this will cause attempts to _send_ on the broken stream to now return `io.ErrPipeClosed` instead of `io.EOF`.

In order to control the error received by the write half (senders), you must close the read half (which is what it was doing before), and vice versa.

Reply:
`Close` is just calling `CloseWithError(nil)` which then gets set to `io.EOF`: https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/io/pipe.go;l=100
It's always the opposite side of the pipe that gets the error with the current side getting: `io.ErrClosedPipe`.

Reply:
`SetError` called close on read: https://github.com/connectrpc/connect-go/blob/d9b61e93e4e7e8c2a4c6be7a31fee7c33e7dc9a3/duplex_http_call.go#L236
Was my mistake switching to `Writer`, it's now fixed in the latest PR.

Reply:
> the current side getting: `io.ErrClosedPipe`
> `SetError` called close on read:

Exactly. By closing the writer here, we will induce `io.ErrClosedPipe` in calls to send (which is what used the writer). The old behavior, closing the reader, is better since it results EOF instead.

> Was my mistake switching to Writer, it's now fixed in the latest PR.

I don't follow this. I'm looking at refreshed diff and it's still closing the writer.

Reply:
Sorry, I confused myself and switched the ordering. My previous comments don't align.

Closing the write side makes `Write` on the pipe error with `io.ErrClosedPipe`. Connect checks for this when sending: https://github.com/connectrpc/connect-go/blob/d9b61e93e4e7e8c2a4c6be7a31fee7c33e7dc9a3/duplex_http_call.go#L106
However, on the read side we need to return `io.EOF` to avoid HTTP2 setting the `io.ErrClosedPipe` as the error for the response body: https://github.com/connectrpc/connect-go/pull/594#discussion_r1372148665

Reply:
`CloseWrite` behaviour was different to `SetError`: https://github.com/connectrpc/connect-go/blob/d9b61e93e4e7e8c2a4c6be7a31fee7c33e7dc9a3/duplex_http_call.go#L133
https://github.com/connectrpc/connect-go/blob/d9b61e93e4e7e8c2a4c6be7a31fee7c33e7dc9a3/duplex_http_call.go#L236

Reply:
Got it. Thank you for explaining!